### PR TITLE
Raw Streams Garbage Collection

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
Made all raw streams (i.e. streams that are not consumed by the VLO)
thread-safe and extended the garbage collection process to include
the CorfuRuntime::StreamView.

Why should this be merged: Fixes memory leaks created by streams that don't belong to SMRObjects (i.e. transaction stream and any stream created by the CorfuRuntime consumer directly). 

Related issue(s) (if applicable): #1573

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/ViewsGarbageCollector.java [hunk 1]
```diff
@@ -75,6 +75,7 @@ public void runRuntimeGC() {
                     currTrimMark);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/ViewsGarbageCollector.java]: getTrimMark is not a safe-call remove Unrecoverable exceptions from its call path.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 2]
```diff
@@ -471,6 +471,7 @@ private void initializeHeadAndTails() {
     private void cleanUpForRetry() {
         runtime.getAddressSpaceView().invalidateClientCache();
         runtime.getObjectsView().getObjectCache().clear();
+        runtime.getStreamsView().getStreamCache().clear();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Why is this needed? the fast loader only creates unsafe streams.

File: runtime/src/main/java/org/corfudb/runtime/ViewsGarbageCollector.java (hunk 1 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/ViewsGarbageCollector.java]: Done.

File: test/src/test/java/org/corfudb/runtime/view/stream/BackpointerStreamViewTest.java [hunk 3]
```diff
@@ -158,7 +158,7 @@ public void headOfStreamBackpointerTermination() {
 
         // Fetch Stream B and verify backpointer count (which requires 1 read = 1 entry)
         svB.remainingUpTo(totalEntries);
-        assertThat(((BackpointerStreamView) svB).getBackpointerCount()).isEqualTo(1L);
+        assertThat(((ThreadSafeStreamView) svB).getUnderlyingStream().getBackpointerCount()).isEqualTo(1L);
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/runtime/view/stream/BackpointerStreamViewTest.java]: Add an explicit test.

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 4]
```diff
@@ -272,11 +273,10 @@ public Token getTrimMark() {
                             .flatMap(stripe -> stripe.getLogServers().stream())
                             .map(e::getLogUnitClient)
                             .map(LogUnitClient::getTrimMark)
-                            .map(CFUtils::getUninterruptibly)
+                            .map(CompletableFuture::join)
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: unrecoverable exceptions are escaping the try/catch logic and causing some threads to unexpectedly disappear. Don't want to change the getUninterruptibly logic in this PR.

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java [hunk 5]
```diff
@@ -147,7 +148,7 @@ private Layout getLayoutUninterruptibly() {
                     WrongEpochException we = (WrongEpochException) re;
                     log.warn("Got a wrong epoch exception, updating epoch to {} and "
                             + "invalidate view", we.getCorrectEpoch());
-                } else if (re instanceof NetworkException) {
+                } else if (re instanceof NetworkException || re.getCause() instanceof NetworkException) {
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: Its wrapped by CompletionException, I think the java streaming logic wraps it, I'll take a look.
Also, need to retry on TimeoutException.

File: test/src/test/java/org/corfudb/runtime/view/stream/BackpointerStreamViewTest.java (hunk 3 above)
Comment (Maithem) [reply test/src/test/java/org/corfudb/runtime/view/stream/BackpointerStreamViewTest.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java [hunk 6]
```diff
@@ -50,15 +58,37 @@ public IStreamView get(UUID stream) {
         return this.get(stream, StreamOptions.DEFAULT);
     }
 
+    public IStreamView getUnsafe(UUID stream) {
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java]: Can we add a comment on how getUnsafe differentiates from get.. maybe implications/risks of using one versus the other or why we use this for streams consumed by the VLO.

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java [hunk 7]
```diff
@@ -145,7 +145,7 @@ public CorfuCompileProxy(CorfuRuntime rt, UUID streamID, Class<T> type, Object[]
         this.serializer = serializer;
 
         underlyingObject = new VersionLockedObject<T>(this::getNewInstance,
-                new StreamViewSMRAdapter(rt, rt.getStreamsView().get(streamID)),
+                new StreamViewSMRAdapter(rt, rt.getStreamsView().getUnsafe(streamID)),
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: thinking, we could add a comment on why we do not need a thread-safe call on the streams view here...

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java [hunk 8]
```diff
@@ -1,6 +1,7 @@
 package org.corfudb.runtime.view;
 
 import java.time.Duration;
+import java.util.concurrent.CompletionException;
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: is this needed?

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 9]
```diff
@@ -15,6 +15,9 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: is ExecutionException needed?

File: runtime/src/main/java/org/corfudb/runtime/view/stream/ThreadSafeStreamView.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.StreamOptions;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ * A thread-safe implementation of IStreamView.
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/ThreadSafeStreamView.java]: Can we add a comment on which methods are made thread-safe with regards to BackpointerStreamView, it would ease comparison of both.

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 7 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java (hunk 9 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: removed.

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java (hunk 8 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: done

File: runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java (hunk 6 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/view/stream/ThreadSafeStreamView.java (hunk 10 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/stream/ThreadSafeStreamView.java]: BackpointerStreamView is hidden, the IStreamView interface is thread-safe (i.e. all methods).

**Comment by Maithem** [comment]
@annym I see that you have re-enabled the cron (after) i have disabled it. Is it possible to prioritize this?

**Comment by zalokhan** [comment]
@Maithem triggered injection framework.

**Comment by Maithem** [comment]
Thanks for pushing this through guys.
